### ISSUE TITLE
Arbitrary composition of ParallelIterators?

### ISSUE THREAD

**Issue by valarauca** [rayon-rs/rayon#753]
Hello - I was wondering what the preferred way is to compose arbitrary `ParallelIterator` (like for instance `Box<dyn ParallelIterator<Item=T>>`) without creating a lot of shortly lived collections, or removing & re-adding data to threadpool constantly.

I am currently working on some code which takes a user input, and converts it into a final iterator. This functions mostly by passing around data types which take the format of:

```rust
pub type Iter<T> = Box<dyn Iterator<Item=T> +'static>;
pub type Init<T>  = Box<dyn Fn() -> Iter<T> +'static>;
pub type Combinator<T> = Box<dyn Fn(Iter<T>,Iter<T>) -> Iter<T> +'static>;
pub type Coalesce<T> = Box<Fn(Iter<T>) -> Init<T> +'static;
```
[actual code (pardon me, it is a messy personal project)](https://github.com/valarauca/dice_backend/blob/4b16b610599fe43f1586d39cd56f54cbd92a04a2/dice_backend/src/ordering/lambda.rs#L11-#L38)

Now I would prefer to use `ParallelIterator` as the obvious advantages of parallelism (for larger workloads). The problem is that `ParallelIterator` requires ` Sized` (as does almost every trait within [`ranyon::iter:plumbing`](https://github.com/rayon-rs/rayon/blob/44b641e7a8b664c47c35109195441a1e54f1c1b9/src/iter/plumbing/mod.rs) so I can't create a `Box<dyn ParallelIterator>` which makes the upgrade a little more _complicate_.

While I understand I can just synthesize a collection after every stage of iteration, judging how [`collect_extended`](https://github.com/rayon-rs/rayon/blob/44b641e7a8b664c47c35109195441a1e54f1c1b9/src/iter/from_par_iter.rs#L11) is implemented this would require _removing_ that data from the individual threadpools, which seems like a net loss. As not only do I lose most the advantage of parallelism, but I also need _many_ short term big heap collections.

**TL;DR**: How to avoid creating intermediate collections, and just pass around something like `Box<dyn Iterator>`? Or should I just focus on "_building one big lambda_" to give to Rayon?

**Comment by valarauca** [comment]
Label: Question (I can't set it)

**Comment by cuviper** [comment]
You're right that we don't support `dyn Trait` objects in rayon -- this was also discussed in #628.

Can you express these in terms of `dyn Fn`, but with concrete iterator types?

> Or should I just focus on "_building one big lambda_" to give to Rayon?

If you can do so, that sounds like it would be less painful.

**Comment by valarauca** [comment]
Disappointing b/c the model of just "_chaining together a bunch of arbitrary iterators_" under pins the SQL Volcano model of execution (and to a degree optimization, pushing iterators & predicates down the execution tree recursively).

Thanks fro the feedback. Have a great week!
